### PR TITLE
"Add members" user interface / route

### PR DESCRIPTION
**Pull request by bboro**
This builds on top of PR #418 to address issue #417.

I added some test coverage as requested on https://github.com/Gizra/og/pull/418#issuecomment-686398776

UI Changes:

### Local task screenshot

image

### Form

image

### CODE REVIEW

**Comment by MPParsley** [comment]
@bboro, could you add a screenshot of the local task and the form?

**Comment by bboro** [comment]
### Local task screenshot

image

### Form

image

I think currently the "Roles" select still doesn't work - has absolutely 0 effect, so you have to add the role after you add the user. On one of our production sites, we've disabled the "Roles" select with a form alter. I've tried to suggest a fix to assign the roles in the submit handler, however apparently it "should" work without the extra code in the submit handler since it's one of the entity's properties, so there may be deeper issues there :shrug: 

Edit: Adding screenshots to PR summary

**Comment by MPParsley** [comment]
```
FILE: /home/travis/build/Gizra/og/tests/src/Unit/OgAdminRoutesControllerTest.php
--------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
--------------------------------------------------------------------------------
 225 | ERROR | [x] Expected 1 space after FUNCTION keyword; 0 found
 229 | ERROR | [x] Expected 1 space after FOREACH keyword; 0 found
--------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
--------------------------------------------------------------------------------
FILE: /home/travis/build/Gizra/og/tests/src/Functional/AddMemberTest.php
--------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------
 76 | ERROR | The array declaration extends to column 91 (the limit is 80). The
    |       | array content should be split up over multiple lines
--------------------------------------------------------------------------------
```

File: tests/src/Unit/OgAdminRoutesControllerTest.php [hunk 1]
```diff
@@ -202,11 +214,28 @@ public function testRoutesWithAccess() {
...
+   * @covers ::overview
+   */
+  public function testHiddenOverviewRoutesAreFiltered() {
+    $result = $this->getRenderElementResult(TRUE);
+    // Filter all hidden routes.
+    $hidden_routes_info = array_filter($this->routesInfo, function($route) {
+      return isset($route['_overview']) && $route['_overview'] === FALSE;
+    });
+    // Ensure the hidden routes are not returned in the rendered element result.
+    foreach($hidden_routes_info as $key => $route_info) {
```
Comment (MPParsley) [inline tests/src/Unit/OgAdminRoutesControllerTest.php]: Expected 1 space after FOREACH keyword; 0 found:

```suggestion
    foreach ($hidden_routes_info as $key => $route_info) {
```

File: tests/src/Functional/AddMemberTest.php [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  protected $nonMember;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
+    parent::setUp();
+    // Create bundle.
+    $this->groupBundle = mb_strtolower($this->randomMachineName());
+    $bundle = NodeType::create(['type' => $this->groupBundle, 'name' => $this->groupBundle]);
```
Comment (MPParsley) [inline tests/src/Functional/AddMemberTest.php]: The array declaration extends to column 91 (the limit is 80). The array content should be split up over multiple lines:

```suggestion
    $bundle = NodeType::create([
      'type' => $this->groupBundle,
      'name' => $this->groupBundle,
    ]);
```

File: tests/src/Unit/OgAdminRoutesControllerTest.php [hunk 3]
```diff
@@ -202,11 +214,28 @@ public function testRoutesWithAccess() {
...
 
+  /**
+   * Tests overview with hidden routes.
+   *
+   * @covers ::overview
+   */
+  public function testHiddenOverviewRoutesAreFiltered() {
+    $result = $this->getRenderElementResult(TRUE);
+    // Filter all hidden routes.
+    $hidden_routes_info = array_filter($this->routesInfo, function($route) {
```
Comment (MPParsley) [inline tests/src/Unit/OgAdminRoutesControllerTest.php]: Expected 1 space after FUNCTION keyword:

```suggestion
    $hidden_routes_info = array_filter($this->routesInfo, function ($route) {
```

**Comment by MPParsley** [comment]
Hi @bboro, thank you for the screenshots. This is partially the same functionality as provided by #231 ([see screens](https://github.com/Gizra/og/files/772904/membership_ui.pdf)) although somewhat lighter (e.g. there's no membership type, no edit screen).

@amitaibu, @pfrenssen, we should decide what direction to go here.

File: src/Form/GroupSubscribeForm.php [hunk 4]
```diff
@@ -90,7 +90,7 @@ public function getQuestion() {
 
     $message = $this->isStateActive()
       ? $this->t('Are you sure you want to join the group %label?', ['%label' => $label])
-      : $this->t('Are you sure you want to request subscription the group %label?', ['%label' => $label]);
+      : $this->t('Are you sure you want to request a subscription to the group %label?', ['%label' => $label]);
```
Comment (MPParsley) [inline src/Form/GroupSubscribeForm.php]: Fixing this in #715.

**Comment by bboro** [comment]
Looks like #231 is getting a lot of movement lately. I'm assuming that's where the focus is going into if anyone comes in here looking for this functionality. It also fixes the og membership create form's Role assignments too.

**Comment by MPParsley** [comment]
Closing in favor of #231.

**Pull request closed without merging**